### PR TITLE
v2: Ensure grpc outgoing context has metadata

### LINKED ISSUE #4565: v2: grpc headers not passed to model inference
Model metadata from a grpc request is not passed onwards to model.

Creation of a new request id if not existing should be done earlier if needed so it can be passed onwards in headers. This may affect http as well.

### PR DESCRIPTION
**Pull request by cliveseldon**
Fixes #4565 

- Ensure a requestId is present (created or retrieved) before sending grpc request to server
- Ensures outgoing context for proxy grpc call has metadata from incoming context metadata

### CODE REVIEW

File: scheduler/pkg/agent/rproxy_grpc.go [hunk 1]
```diff
@@ -235,22 +235,27 @@ func (rp *reverseGRPCProxy) ModelInfer(ctx context.Context, r *v2.ModelInferRequ
...
-		if errTrailer != nil {
-			logger.WithError(errTrailer).Error("Failed to set trailers")
-		}
+	if trailer == nil {
+		trailer = metadata.MD{}
+	}
+	trailer.Set(util.RequestIdHeader, requestId)
+	errTrailer := grpc.SetTrailer(ctx, trailer) // pass on any trailers set by inference server such as MLServer
+	if errTrailer != nil {
+		logger.WithError(errTrailer).Error("Failed to set trailers")
```
Comment (sakoush) [inline scheduler/pkg/agent/rproxy_grpc.go:258]: could we extract this logic as separate helper?

File: scheduler/pkg/agent/rproxy_grpc.go [hunk 2]
```diff
@@ -235,22 +235,27 @@ func (rp *reverseGRPCProxy) ModelInfer(ctx context.Context, r *v2.ModelInferRequ
 		go rp.metrics.AddModelInferMetrics(externalModelName, internalModelName, metrics.MethodTypeGrpc, elapsedTime, codes.NotFound.String())
 		return nil, status.Error(codes.NotFound, fmt.Sprintf("Model %s not found (err: %s)", r.ModelName, err))
 	}
+
+	// Create an outgoing context for the proxy call to service from incoming context
+	outgoingCtx, requestId := rp.createOutgoingCtxWithRequestId(ctx)
```
Comment (sakoush) [inline scheduler/pkg/agent/rproxy_grpc.go:240]: do we need to do something similar to `ModelMetadata` and `ModelReady` grpc calls?

File: scheduler/pkg/agent/rproxy_grpc_test.go [hunk 3]
```diff
@@ -316,12 +299,13 @@ func TestAddRequestIdToTrailer(t *testing.T) {
 			rp := reverseGRPCProxy{
 				logger: log.New(),
 			}
-			rp.addRequestIdToTrailer(test.ctx, test.trailer)
-			data := test.trailer.Get(util.RequestIdHeader)
```
Comment (sakoush) [inline scheduler/pkg/agent/rproxy_grpc_test.go:320]: As we are removing the trailer tests? could we add them somewhere else?

**Review by sakoush — COMMENTED** [review]
Left some minor comments

File: scheduler/pkg/agent/rproxy_grpc.go (hunk 2 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/rproxy_grpc.go:240]: Probably need to decide if requestId is needed for these calls. Maybe a follow on or when we decide its needed.

File: scheduler/pkg/agent/rproxy_grpc_test.go (hunk 3 above)
Comment (cliveseldon) [reply scheduler/pkg/agent/rproxy_grpc_test.go:320]: The function had been removed

**Review by sakoush — APPROVED** [review]

**Pull request merged**